ambry-api/src/main/java/com.github.ambry/router/AsyncWritableChannel.java

```diff
@@ -44,5 +47,59 @@
    * @return a {@link Future} that will eventually contain the result of the write operation (the number of bytes
    *         written).
    */
-  public Future<Long> write(ByteBuffer src, Callback<Long> callback);
+  Future<Long> write(ByteBuffer src, Callback<Long> callback);
+
+  /**
+   * Write data in {@code src} to the channel and the {@code callback} will be invoked once the write succeeds or fails.
+   * This method is the counterpart for {@link ByteBuf}. It shares the same guarantee as the {@link #write(ByteBuffer, Callback)}.
+   * Whoever implements this interface, shouldn't release the {@code src}. If releasing is expected after finishing writing
+   * to channel, please releas this {@link ByteBuf} in the callback method.
+   * @param src The data taht needs to be written to the channel.
+   * @param callback The {@link Callback} that will be invoked once the write succeeds/fails. This can be null.
+   * @return a {@link Future} that will eventually contain the result of the write operation (the number of bytes
+   *         written).
+   */
+  default Future<Long> write(ByteBuf src, Callback<Long> callback) {
+    int numBuffers = src.nioBufferCount();
+    if (numBuffers == 1) {
+      return write(src.nioBuffer(), (result, exception) -> {
+        if (result != 0) {
+          src.readerIndex(src.readerIndex() + (int) result.longValue());
+        }
+        callback.onCompletion(result, exception);
+      });
+    } else {
+      ByteBuffer[] buffers = src.nioBuffers();
+      AtomicLong size = new AtomicLong(0);
+      AtomicInteger index = new AtomicInteger(0);
+      FutureResult<Long> futureResult = new FutureResult<Long>();
+      Callback<Long> cb = new Callback<Long>() {
+        @Override
+        public void onCompletion(Long result, Exception exception) {
+          index.addAndGet(1);
+          size.addAndGet(result);
+          if (result != 0) {
+            src.readerIndex(src.readerIndex() + (int) result.longValue());
+          }
+          if (exception != null) {
+            futureResult.done(size.get(), exception);
```

Comment:
this could result in the callback and futureResult being called multiple times if multiple chunks got an exception.

How about:
```java
        if (exception != null || index.get() == buffers.length) {
          if (callbackInvoked.compareAndSet(false, true)) {
            futureResult.done(size.get(), exception);
            if (callback != null) {
              callback.onCompletion(size.get(), exception);
            }
          }
        }
```

Reply:
this is better, will update.